### DISCUSSION TITLE
formatter "path.filename_first" needs different filter than display

### DISCUSSION THREAD

**Discussion by Grueslayer** [ibhagwan/fzf-lua#1247] [General]
Hi,

I can not get this running like (I think) it should work.

Let's assume you've following files

```
├───dir1
│   ├───doc
│   │       file1.txt
│   │
│   └───src
│           file1.txt
│
└───dir2
    ├───doc
    │       file1.txt
    │
    └───src
            file1.txt
```

and you go with files and type `srcfile1` you get an empty list intead all file1.txt in all src subfolders (excluding the file1.txt in the doc subdirs)

To make this run the "right" way I think you have to make `fd` ( `rg` ) creating three columns:

1. filename
2. path
3. path+filename (like normal)

and  use `--with-nth = 1,2` and `--nth = 3` so 1+2 are displayed and 3 is filtered.

**Comment by ibhagwan** [comment]
You're correct, currently the format of the string is "backwards", file first and then the folder which results in the behavior you're describing.

Althought what you're desribing is desireable I'd have to check if this is possible as you need a delimiter which isn't part of anywhere else in the string to make this work, you can't use a path separator or `:` as this will break in  many cases (not limted to Windows).

**Comment by ibhagwan** [comment]
This doesn't seem possible with fzf as `--with-nth` doesn't preserve the match order.

```sh
echo "a/b/c/d.txt" | fzf
```
![image](https://github.com/ibhagwan/fzf-lua/assets/59988195/3a3dfcaf-2116-48e8-bbe7-5bd9a58375fd)

Now with:
```sh
 echo "a/b/c/d.txt" | fzf --delimiter="/" --with-nth="4,1..-2"
```
![image](https://github.com/ibhagwan/fzf-lua/assets/59988195/4a6a481a-cb96-4d69-a30c-148ac4247aaa)
> When typing the same text `adtxt` no results are found:
![image](https://github.com/ibhagwan/fzf-lua/assets/59988195/0ae2d869-7349-4e11-9fec-cb3145c319b3)

**Comment by Grueslayer** [comment]
I am thinking about something like this:

```
echo f.txt;a/b/c/e;a/b/c/e/f.txt; | fzf --delimiter ";" --nth 3 --with-nth "1..2"
```

BUT it seems `--nth` only supports fields included in `--with-nth` (https://github.com/junegunn/fzf/issues/1440) so it doesn't work, too.

**Reply by ibhagwan** [reply]
> I am thinking about something like this:
> 
> ```
> echo f.txt;a/b/c/e;a/b/c/e/f.txt; | fzf --delimiter ";" --nth 3 --with-nth "1..2"
> ```
> 
> BUT it seems `--nth` only supports fields included in `--with-nth` ([junegunn/fzf#1440](https://github.com/junegunn/fzf/issues/1440)) so it doesn't work, too.

Yap I tried that as well.

**Comment by ibhagwan** [comment]
@Grueslayer, FYI, see https://github.com/junegunn/fzf/issues/3854, hopefully the maintainer will consider this scenario.

**Reply by ibhagwan** [reply]
Doesn’t look like it’s gonna happen: https://github.com/junegunn/fzf/issues/3854#issuecomment-2159543403.

**Comment by ibhagwan** [comment]
@Grueslayer,  with the latest commit you can now use the experimental v2 of the formatter:
```lua
:lua require("fzf-lua").files({ formatter={"path.filename_first",2} })
-- or
:FzfLua files formatter={"path.filename_first",2}
-- via setup, applied to all
:lua require("fzf-lua").setup({ defaults = { formatter={"path.filename_first",2} } })
```

Ref: https://github.com/ibhagwan/fzf-lua/pull/1255#issuecomment-2163449012

**Reply by gugahoi** [reply]
Thanks for adding this. Just want to check if this is something that you intend to support ongoing or if it will remain as experimental only? I much prefer this way of searching to the default `path.filename_first`.

**Reply by ibhagwan** [reply]
> Thanks for adding this. Just want to check if this is something that you intend to support ongoing or if it will remain as experimental only? I much prefer this way of searching to the default `path.filename_first`.

Currently we’re limited by the upstream, if the upstream adds features making this more native it we’ll adjust.

**Comment by Grueslayer** [comment]
@ibhagwan that seem to work much better. Highlighting of the chars matching the fuzzy word does not work if it is part of path and filename. But I think this isn’t possible without implementing it directly in fzf.

**Reply by ibhagwan** [reply]
> But I think this isn’t possible without implementing it directly in fzf.

You’re correct, this is somewhat of a hack and isn’t solveable without upstream changes.

> could you add the formatter to „history“, too?

Not sure what you mean, this works on all pickers where paths are constructed using `make_entry.file`
```lua
:FzfLua oldfiles formatter={"path.filename_first",2}
```
